miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/permission/MiniAppCustomPermissionCache.kt

```diff
@@ -80,22 +80,44 @@ internal class MiniAppCustomPermissionCache(context: Context) {
      * Stores the grant results to SharedPreferences.
      * @param [miniAppCustomPermission] an object to contain the results per MiniApp.
      */
-    @SuppressLint("RestrictedApi")
     fun storePermissions(
         miniAppCustomPermission: MiniAppCustomPermission
     ) {
         val supplied = miniAppCustomPermission.pairValues.toMutableList()
-
         // Remove any unknown permission parameter from HostApp.
         supplied.removeAll { (first) ->
             first.type == MiniAppCustomPermissionType.UNKNOWN.type
         }
-
         val miniAppId = miniAppCustomPermission.miniAppId
         val allPermissions = prepareAllPermissionsToStore(miniAppId, supplied)
         applyStoringPermissions(MiniAppCustomPermission(miniAppId, allPermissions))
     }
 
+    fun updateManifestPermissions(
+        miniAppId: String,
+        manifestPermissions: List<Pair<MiniAppCustomPermissionType, MiniAppCustomPermissionResult>>
+    ) {
+        val cachedPermissions = readPermissions(miniAppId).pairValues.toMutableList()
+        val changedPermissions = (manifestPermissions + cachedPermissions).groupBy { it.first.type }
+            .filter { it.value.size == 1 }
+            .flatMap { it.value }
+
+        if (changedPermissions.isNotEmpty()) {
+            if (cachedPermissions.size < manifestPermissions.size) {
+                val filteredValue =
+                    MiniAppCustomPermission(miniAppId, cachedPermissions + changedPermissions)
+                applyStoringPermissions(filteredValue)
+            } else {
+                cachedPermissions.removeAll { (first) ->
+                    first.type in changedPermissions.groupBy { it.first.type }
+                }
+
+                val filteredValue = MiniAppCustomPermission(miniAppId, cachedPermissions)
+                applyStoringPermissions(filteredValue)
+            }
+        }
```

Comment:
I'm not really understanding the logic in this function, but maybe I'm missing something. Is this also adding permissions from the manifest to the cached permissions?

Basically we only want to make sure that the cached permissions don't contain any permissions which aren't in the manifest (we don't need to worry about adding any permissions from the manifest to the cached permissions). So couldn't you just do something like this?

```suggestion
        val cachedPermissions = readPermissions(miniAppId).pairValues.toMutableList()
        val newPermissions = cachedPermissions.mapNotNull { cached -> 
            manifestPermissions.find { it.first == cached.first } 
        }
        val filteredValue =
            MiniAppCustomPermission(miniAppId, newPermissions)
        applyStoringPermissions(filteredValue)
```